### ISSUE TITLE
Transaction commit don't seem to work anymore from 95fbc6fac6984b52b115d6c5c51a43ed7cd9ce02

### ISSUE THREAD

**Issue by niclone** [sequelize/sequelize#1158]
Transaction commit don't seem to work anymore after this commit : https://github.com/sequelize/sequelize/commit/95fbc6fac6984b52b115d6c5c51a43ed7cd9ce02

promise resolve on t.commit(), but datas are not commited really.

**Comment by hariprasadkulkarni** [comment]
I'm facing a similar issue. I'm trying to create a record using transactions. And on a commit success, I run a find on the table. The 'find' returns a null response. This happens intermittently. I found this while I was running a load test on my app. I was trying to create 200 users (with say a maximum of 10 being created concurrently). Out of those, around 5 returned a null response when queried using 'find'.

	db.sequelize.transaction({
		"autocommit": false
	}).then(function(trx) {

		var options = {};
		options["transaction"] = trx;

		db.Users
		.create(userAttributes, options)
		.then(function(userObj){

			trx.commit().success(function(){

				getUserById(userObj.id)
				.then(function(success){
					response.send(success);
				},function(error){
					response.send(error);
				});

			});
		})
	})

The getUserById method makes a db.Users.find for the id that is passed. For some calls, the response object is null. When I query the particular user using SQL, I see that the user has been created. I'm using mysql.

**Comment by mickhansen** [comment]
@hariprasadkulkarni What version of sequelize are you using?

**Comment by hariprasadkulkarni** [comment]
@mickhansen I am using 2.0.0-rc8.

**Comment by mickhansen** [comment]
@hariprasadkulkarni I'd start by upgrading then, transactions work way better now.

**Comment by hariprasadkulkarni** [comment]
@mickhansen Hmm. That is quite an effort considering the 3.0.0 changes. I'll try to get it done and let you know.

**Comment by mickhansen** [comment]
@hariprasadkulkarni Well you could upgrade to 2.0 atleast ;) But we can't really say what's wrong with that old a version, but earlier dev versions of 2.0 had major issues with transactions.

**Comment by hariprasadkulkarni** [comment]
@mickhansen Oh of course. :) It's just that in case you are waiting on me for a reply on this, you may have to wait for a few weeks. But I will surely get back to you on this.

**Comment by alockwood05** [comment]
@mickhansen I'm seeing the issue on the latest version of sequelize 3.20.  in some tests I create data in transactions, three items, and the third item doesn't correctly resolve.  If I add the code 
`.then(() => db.transaction().then(tx => tx.commit()))`  then everything works as expected... it seems like the commit doesn't actually return the commit promise resolution...

**Comment by mickhansen** [comment]
@alockwood05 This is an ancient issue so i doubt you're facing the issues described here.

**Comment by alockwood05** [comment]
Ok I'll get a test to repro or disregard, thanks for quick response.

**Comment by alockwood05** [comment]
@mickhansen - I was implementing seed data for a school database (institution, branch) and the transaction when resolved didn't work as expected.   should this be a new bug or does it help to have on this already existing thread?

This is my specific example: 
```
export const SCHOOL_FIXTURES = [
  {
    name: 'A College',
    opeid6: '000001',
    opeid8: '00000101',
    ipeds: '100001',
    address_line1: '1 street',
    city: 'Chestnut Hill',
    state_abbr: 'MA',
    country: 'USA',
    additionalBranches: [{
      name: 'A College Other Location',
      opeid6: '000002',
      opeid8: '00000202',
      ipeds: '100002',
      address_line1: '2 street',
      city: 'Other City',
      state_abbr: 'MA',
      country: 'USA'
    }]
  },
  {
    name: 'Harvard',
    opeid6: '000003',
    opeid8: '00000301',
    ipeds: '100003',
    address_line1: '3 my. ,!@# street',
    city: 'Chestnut Hill',
    state_abbr: 'MA',
    country: 'USA'
  },
  {
    name: 'SOME COLLEGE',
    opeid6: '000004',
    opeid8: '00000401',
    ipeds: '100004',
    address_line1: '4 street',
    city: 'OTHER Town',
    state_abbr: 'CA',
    country: 'USA'
  }
];
Object.freeze(SCHOOL_FIXTURES);
export function persistSchoolFixture(db, models, _schoolData) {
  const schoolData = _.cloneDeep(_schoolData);
  // tmp so that the notnull constraint doesn't fail
  schoolData.canonical_branch_id = 0;
  const branches = [];
  branches.push(schoolData);
  if (schoolData.additionalBranches) {
    Array.prototype.push.apply(branches, schoolData.additionalBranches);
  }
  const institutionData = Object.assign({}, schoolData, {branches});
  // wrap it all in a dbx
  return db.transaction().then(tx => {
    models.Institution.create(institutionData, {
      transaction: tx, include:[{model: models.Branch, as: 'branches'}]
    })
    .then(institution => {
      const canonicalBranch = institution.branches.find(item => item.ipeds === institution.ipeds);
      return institution.setCanonicalBranch(canonicalBranch, {transaction: tx});
    })
    .then(result => tx.commit().return(result))
    .catch(err => tx.rollback().throw(err));
  }).then(function () {
    // THE FindAll Doesn’t return the school branch that was just created . :(
    return models.Branch.findAll().then(branches => branches && branches.map(branches => branches.name)).tap(console.log);
  });
}

// WORKS as expected
// export function persistSchoolFixture(db, models, _schoolData) {
//   const schoolData = _.cloneDeep(_schoolData);
//   const branches = [];
//   branches.push(schoolData);
//   if (schoolData.additionalBranches) {
//     Array.prototype.push.apply(branches, schoolData.additionalBranches);
//   }
//   const institutionData = Object.assign({}, schoolData, {branches});
//   return models.Institution.create(institutionData, {
//     include:[{model: models.Branch, as: 'branches'}]
//   })
//   .then(institution => {
//     const canonicalBranch = institution.branches.find(item => item.ipeds === institution.ipeds);
//     return institution.setCanonicalBranch(canonicalBranch);
//   })
//   .then(function () {
//     // debugging
//     return models.Branch.findAll()
//       .then(branches => branches && branches.map(branches => branches.name)).tap(console.log);
//   });
// }
```

**Comment by mickhansen** [comment]
You're missing a return in your then callback.

**Comment by eurostar-fennec-cooper** [comment]
Hello from the future!

Fyi this does seem to cause issues with deferred constraint checking in Postgres. If I defer a unique constraint check to commit time then it's possible that PG will respond with an error to the commit statement. In which case the commit promise does not seem to be rejected, but resolved instead. Nor does the query within the transaction that caused it throw either (because the constraint was deferred).

Was crawling up the stack in the debugger to see where it's just giving up and resolving but I couldn't easily spot it.
